src/hamcrest/core/core/future.py

```diff
@@ -0,0 +1,138 @@
+            return False
+
+        exc = future.exception()
+        if exc is None:
+            return False
+
+        if isinstance(exc, self.expected):
+            if self.pattern is not None:
+                if re.search(self.pattern, str(exc)) is None:
+                    return False
+            if self.matcher is not None:
+                if not self.matcher.matches(exc):
+                    return False
+            return True
+
+        return False
+
+    def describe_to(self, description: Description) -> None:
+        description.append_text("Expected a future with exception %s" % self.expected)
+
+    def describe_mismatch(self, future: asyncio.Future, description: Description) -> None:
+        if not asyncio.isfuture(future):
+            description.append_text("%s is not a future" % future)
+            return
+
+        if not future.done():
+            description.append_text("%s is not completed yet" % future)
+            return
+
+        if future.cancelled():
+            description.append_text("%s is cancelled" % future)
+            return
+
+        exc = future.exception()
+        if exc is None:
+            description.append_text("No exception raised.")
+        elif isinstance(exc, self.expected):
+            if self.pattern is not None or self.matcher is not None:
+                description.append_text("Correct assertion type raised, but ")
+                if self.pattern is not None:
+                    description.append_text('the expected pattern ("%s") ' % self.pattern)
+                if self.pattern is not None and self.matcher is not None:
+                    description.append_text("and ")
+                if self.matcher is not None:
+                    description.append_description_of(self.matcher)
+                    description.append_text(" ")
+                description.append_text('not found. Exception message was: "%s"' % str(exc))
+        else:
+            description.append_text("%r of type %s was raised instead" % (exc, type(exc)))
+
+    def describe_match(self, future: asyncio.Future, match_description: Description) -> None:
+        exc = future.exception()
+        match_description.append_text("%r of type %s was raised." % (exc, type(exc)))
+
+
+def future_exception(
+    exception: Type[Exception], pattern=None, matching=None
+) -> Matcher[asyncio.Future]:
+    """Matches a future with the expected exception.
+
+    :param exception:  The class of the expected exception
+    :param pattern:    Optional regular expression to match exception message.
+    :param matching:   Optional Hamcrest matchers to apply to the exception.
+
+    Expects the actual to be an already resolved future.  Optional argument
+    pattern should be a string containing a regular expression.  If provided,
+    the string representation of the actual exception - e.g. `str(actual)` -
+    must match pattern.
+
+    Examples::
+
+        assert_that(somefuture, future_exception(ValueError))
```

Comment:
Should we be leaning in to `assert_that()` here, or conceiving of a different assert wrapper?

Reply:
I thought about creating a variant that itself could by `awaited` but I think that would be much more complicated and impact how matcher and everything works. This is really the huge problem with asyncio (and async code in general) it tends to just infect everything and spread to all corners of your code.